### PR TITLE
Document why we placed ELBs in front of everything

### PR DESCRIPTION
**Pull request by bleach**
I picked ADR007 because it was available and roughly the correct cronology.

### CODE REVIEW

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+- Deployed applications (gorouter). Accessed by the public.
+- CF API. Accessed by tenants.
+- UAA. Accessed by tenants.
+- Loggregator. Accessed by tenants.
+- SSH proxy. In theory accessed by tenants, but not working in our environment.
+
+We had an existing credentials store suitable for storing the private keys at
+rest. Only a small number of engineers within the team can access; the same
+ones that can make IAM changes using our account=wide terraform config.
```
Comment (mtekel) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: typo: account-wide

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  loggregator, which relies on them for log streaming. It would also prevent
+  tenants from using web sockets within their applications.
+- When the ELB is in TCP mode, we have no way of communicating the client IP
+  address to the downstream service. Practical consequences of this would be
+  that tenants would be unable to see in their logs who is using their service or
+  do any access control based on client IP address.
+
+In attempting to solve the second problem, we explored some options:
+
+- ELB has support for the [Proxy Protocol], but unfortunately none of the
```
Comment (alext) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: It looks like you intended to add a link here for "Proxy Protocol".

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+Consequences
+============
+
+- We played a [spike to investigate setting X-Forwarded-For
+  correctly](https://www.pivotaltracker.com/projects/1275640/stories/116619465)
+which produced an [upstream PR to add proxy protocol support to
+gorouter](https://github.com/cloudfoundry/gorouter/pull/126)
+- We also added a way to introduce X-Forwarded-Proto headers [PR](https://github.com/cloudfoundry/gorouter/pull/127)
+- As an interim measure until gorouter gained support, [we needed to work on](https://www.pivotaltracker.com/story/show/116309951)
```
Comment (alext) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: This reads as though it's part of the previous bullet-point. Did you mean to include the leading `-`?

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  that tenants would be unable to see in their logs who is using their service or
+  do any access control based on client IP address.
+
+In attempting to solve the second problem, we explored some options:
+
+- ELB has support for the [Proxy Protocol], but unfortunately none of the
+  downstream services eg. gorouter support it. It seemed simple to add support
+  to gorouter.
+- We could introduce another intermediary proxy such as HAProxy, which
+  understands the proxy protocol and adds or appends to an `X=Forwarded=For`
```
Comment (alext) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: Typo: `X-Forwarded-For`

**Comment by alext** [comment]
A couple of minor comments, otherwise it all looks good to me.

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md (hunk 1 above)
Comment (bleach) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: Fixed

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md (hunk 2 above)
Comment (bleach) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: Correct :) Fixed

File: docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md (hunk 3 above)
Comment (bleach) [inline docs/architecture_decision_records/ADR007-terminating-tls-at-elbs.md]: I merged it into the previous bullet and also reworded slightly.

**Comment by bleach** [comment]
Good feedback. I think I've addressed all the comments now.

**Pull request merged**